### PR TITLE
Pipeline: add write once containers

### PR DESCRIPTION
**Pull request by drblallo**

### CODE REVIEW

File: include/revng/Pipeline/Loader.h [hunk 1]
```diff
@@ -211,10 +226,15 @@ class Loader {
...
 template<>
-struct llvm::yaml::MappingTraits<pipeline::ContainerDeclaration>
-  : public TupleLikeMappingTraits<pipeline::ContainerDeclaration> {};
+struct llvm::yaml::MappingTraits<pipeline::ContainerDeclaration> {
+  static void mapping(IO &TheIO, pipeline::ContainerDeclaration &Info) {
+    TheIO.mapRequired("Name", Info.Name);
+    TheIO.mapOptional("Type", Info.Type);
+    TheIO.mapOptional("Role", Info.Role);
+  }
+};
```
Comment (aleclearmind) [inline include/revng/Pipeline/Loader.h:237]: Why?

File: include/revng/Pipeline/RegisterContainerFactory.h [hunk 2]
```diff
@@ -56,4 +56,31 @@ class RegisterDefaultConstructibleContainer : Registry {
   void libraryInitialization() override {}
 };
 
+/// Instantiate a global object of this class
+/// for each default constructible container to be used as role you wish to
+/// register
+template<typename ContainerT>
+class RegisterDefaultConstructibleRole : Registry {
```
Comment (aleclearmind) [inline include/revng/Pipeline/RegisterContainerFactory.h:63]: Let's just have `RegisterRole`, we'll never have non default constructible roles.
It's just a list of strings, let's not make it more complex that it needs to be.

File: lib/Pipeline/Loader.cpp [hunk 3]
```diff
@@ -123,14 +128,40 @@ Loader::parseInvocation(Step &Step, const PipeInvocation &Invocation) const {
...
   auto &Entry = It->second;
+  for (const auto &ContainerName : Invocation.UsedContainers) {
+
+    if (llvm::find(WriteOnceNames, ContainerName) == WriteOnceNames.end())
+      continue;
+
+    if (PipelineContext->hasRegisteredWriteOnceContainer(ContainerName))
+      return createStringError(inconvertibleErrorCode(),
+                               "Detected two uses of write once container %s\n",
+                               ContainerName.c_str());
```
Comment (aleclearmind) [inline lib/Pipeline/Loader.cpp:139]:
```suggestion
    if (PipelineContext->hasRegisteredWriteOnceContainer(ContainerName)) {
      return createStringError(inconvertibleErrorCode(),
                               "Detected two uses of write once container %s\n",
                               ContainerName.c_str());
     }
```

File: include/revng/Pipeline/Context.h [hunk 4]
```diff
@@ -96,6 +101,28 @@ class Context {
 
   const GlobalsMap &getGlobals() const { return Globals; }
 
+  void addWriteOnceContainer(llvm::StringRef Name,
```
Comment (aleclearmind) [inline include/revng/Pipeline/Context.h:104]: Can we come up with a name that's more intuitive than "WriteOnceContainer"?
Maybe "ConstantContainer" or "GlobalContainer" (sad).

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: include/revng/Pipeline/Loader.h (hunk 1 above)
Comment (drblallo) [reply include/revng/Pipeline/Loader.h:237]: before type was mandatory, now it is either type or role

File: include/revng/Pipeline/RegisterContainerFactory.h (hunk 2 above)
Comment (drblallo) [reply include/revng/Pipeline/RegisterContainerFactory.h:63]: mhh i think we will, in particular if a container must be configured with a type

Comment (aleclearmind) [reply include/revng/Pipeline/RegisterContainerFactory.h:63]: That can be a template argument. Why do you think we'll need a constructor?

Comment (drblallo) [reply include/revng/Pipeline/RegisterContainerFactory.h:63]: mistake, i meant a kind

File: include/revng/Pipeline/Loader.h (hunk 1 above)
Comment (drblallo) [reply include/revng/Pipeline/Loader.h:237]: i am doing the current pattern in all other mapping in that file, should i switch over to the suggested pattern in all of them?

File: include/revng/Pipeline/RegisterContainerFactory.h (hunk 2 above)
Comment (aleclearmind) [reply include/revng/Pipeline/RegisterContainerFactory.h:63]: I'd expect a role to be a map from string to container, why do you envision the need for a kind?

File: include/revng/Pipeline/Loader.h (hunk 1 above)
Comment (aleclearmind) [reply include/revng/Pipeline/Loader.h:237]: Consider if it's worth it.
You could also consider switch to tuple trees.

**Pull request closed without merging**